### ISSUE TITLE
DataProcessorUI: make post-processing optional

### ISSUE THREAD

**Issue by raquelalvarezbanos** [mantidproject/mantid#16632]
Make the post-processing step optional. This involves:

- Working without the 'Group' columns (groups are only used to tell the widget what workspaces need to be post-processed together)
- Removing option to warn users when a group is going to be partially processed.

**Comment by raquelalvarezbanos** [comment]
Reflectometry scientists have requested a change in the way the data is displayed, they would like to use a `QTreeView` instead of the current `QTableView`. The reason is that they will have hundreds of different groups at some point, and they would like to hide sub-items, and only inspect them if something goes wrong with the reduction.

![newtreeview](https://cloud.githubusercontent.com/assets/8956985/16263837/4d064c5e-386e-11e6-92cd-71a3a7d10928.png)

This is somehow related to this issue: with the current `QTableView`, when no post-processing is needed, there is no need to show the `Group` column (actually this column should be removed to avoid confusion), whereas if a post-processing algorithm is specified, the `Group` column is necessary. With a new `QTreeView`, we will never have a `Group` column, the only difference between having a post-processing step or not will be that items will be displayed hierarchically (with child items) or as single items with no children.

I'll change the type of view as part of this PR and then I'll make the post-processing optional.

**Comment by raquelalvarezbanos** [comment]
Too many changes for a single issue, I've created a new one to move from `QTableView` to `QTreeView` that should be fixed before this one.

Blocked by:
- [ ] https://github.com/mantidproject/mantid/issues/16734

**Comment by raquelalvarezbanos** [comment]
When no post-processing is specified:

- Some actions do not make sense (`ExpandSelection`, `PlotGroups`, etc) and should not be in the toolbar.
- Actions that still make sense will probably have to behave differently
- Therefore, descriptions will probably be different from the case where a post-processing algorithm is given.

**Comment by raquelalvarezbanos** [comment]
I have been looking into the best way of making post-processing step optional and I am starting to think that perhaps it would be better two have two concrete presenters, each one dealing with a specific case: one for the case where groups are needed and the other one for the case where no post-processing is needed. The reason is that if I keep the existing presenter, the following methods would have to have `if (m_postprocess)` statements (where `m_postprocess` would be a boolean indicating whether post-processing is needed):

- `GenericDataProcessorPresenter::initOptions()` because there are options such as `WarnProcessPartialGroup` that do not make sense when there are no groups.
- `GenericDataProcessorPresenter::validateModel()` because it checks that the number of columns in the table workspace in terms of the number of elements in the whitelist: for post-processing the table workspace has to have one extra column indicating the group (the whitelist is no longer aware of this now that we have moved to a `QTreeView`). When no post-processing is needed both should have the same number of columns.
- `GenericDataProcessorPresenter::createDefaultWorkspace()` for the same reason.
- `GenericDataProcessorPresenter::process()`, because the presenter requests selected groups to the view, which does not make sense if we have groups.
- `GenericDataProcessorPresenter::postProcessGroup()` should not exist if we don't have groups.
- `GenericDataProcessorPresenter::reduceRow()` because to reduce a row we need to access data in the tree via their index, and this is different if we have a two-layer tree or a one-layer tree.
- Some actions (insert group, delete group, expand group, etc) should not exist if there are no groups.
- Actions that do make sense will behave differently if we have a two-layer tree or a one-layer tree.

Methods that can probably be common to both cases:
- `GenericDataProcessorPresenter::prepareRunWorkspace()`
- `GenericDataProcessorPresenter::loadRun()`

**Comment by raquelalvarezbanos** [comment]
To summarize: 

This issue is about making the post-processing step in the `DataProcessorUI` optional:
- If a post-processing algorithm is given, it means that we have groups of runs, which in turn means that our tree (we have recently moved from a `QTableView` to a `QTreeView`) is a 2-level tree with parent items (groups) and child items (rows/runs)
- If no post-processing algorithm is given it means that we don't need groups, which means that a 1-level tree in which items have no children is enough.

Therefore, making post-processing optional involves having two different `QTree` models. This is fine, I've already implemented a 1-layer model and it seems to be working on my local branch.

I see three different options for the presenter:

**Option 1**

We can have our current `GenericDataProcessorPresenter` with a boolean member variable, `m_postprocess`, to indicate whether or not we have to post-process data. This will require perhaps too many `if`/`else` statements, depending on whether or not we have to post-process the data. This happens each time we need to know what items were selected by users and act accordingly, which is the case in almost all of the actions. As an example, this:

```
// Selected groups
groups = m_view->getSelectedParentItems();
// Selected rows
rows = m_view->getSelectedChildItems();
```

should become:

```
if (m_postprocessing) {
      // Selected groups
      groups = m_view->getSelectedParentItems();
      // Selected rows
      rows = m_view->getSelectedChildItems();
} else
{
      // Selected rows
      rows = m_view->getSelectedParentItems();
      // I'll also have to set an invalid index for groups
      // so that when we read/write data from/to the model calls like this:
      // m_model->data(m_model->index(row, col, m_model->index(group, 0)))
      // work
      groups.insert(-1);
}
```

**Option 2**

We could refactor the `GenericDataprocessorPresenter` so that it keeps common functionality (for instance `importTable()`, `exportTable()`, `saveTableAs()`, etc, which do not depend on the type of model), but delegates some work to a new class, `DataProcessorTreeManager` (not sure if this is the best name). This would be an abstract class with two different implementations, `DataProcessorOneLayerTreeManager` and `DataProcessorTwoLayerTreeManager`, so that all we have to do when we create the generic presenter is:

```
GenericDataProcessorPresenter::GenericDataProcessorPresenter(...) {
  if (postprocessing) {
    m_treeManager = make_unique<DataProcessorTwoLayerTreeManager>();
  }
  else {
    m_treeManager = make_unique<DataProcessorOneLayerTreeManager>();
  }
}
```

Then, for actions, we could simply delegate the work like this:

```
GenericDataProcessorPresenter::appendRow() {
  m_treeManager ->appendRow();
}
```

This new class would also be in charge of deciding with options have to be shown on the toolbar and what commands to publish (for instance if there are no groups, `plotGroup()`, `expandSelection()` etc do not make sense). In summary, the generic presenter would deal with algorithms (it would set input properties, run the algorithms, etc) and the "tree manager" would read/write from/to the model.

The only thing to consider with this approach, is that this new class would need to know what items were selected by the user, which is known by the view, so this would have to go through the presenter. In addition, when the manager needs to print any type of message, it would also have to act on the presenter.

**Option 3**

We could also have two completely separate concrete presenters, for instance a `GenericDataProcessorPresenterTwoLevelTree` and a `GenericDataProcessorPresenterOneLevelTree`. I don't like this option that much because I think we would have some functionality duplicated. 

I've tried a first attempt of both options 1 and 2, and I would like your opinion on this, @OwenArnold . Right now, I am leaning more towards option 2, but I'm concerned I could be overcomplicating things.
